doc/dvsource-alsa.1
.\" dvsource-alsa.1 written by Ben Hutchings <ben@decadent.org.uk>
.mso www.tmac
.TH DVSOURCE-DVGRAB 1 "11 March 2009"
.SH NAME
dvsource-alsa \- audio source for DVswitch
.SH SYNOPSIS
.HP
.B dvsource-alsa
.RI [ OPTIONS ]
.RB [ \-s " ntsc|pal]
.RB [ \-r " 48000|32000|44100]
.RI [ DEVICE ]
.SH DESCRIPTION
.LP
\fBdvsource-alsa\fR uses ALSA to capture audio and sends it to
DVswitch together with a dummy video stream.
.LP
The optional \fIDEVICE\fR argument is a PCM device name. See
.URL http://www.alsa-project.org/alsa-doc/alsa-lib/pcm.html#pcm_dev_names "the ALSA documentation"
for an explanation of these names.
.SH OPTIONS
\fB\-h\fR, \fB\-\-host=\fIHOST\fR
.TP
\fB\-p\fR, \fB\-\-port=\fIPORT\fR
.RS
Specify the network address on which DVswitch is listening.  The host
address may be specified by name or as an IPv4 or IPv6 literal.
.RE
.TP
\fB\-s\fR, \fB\-\-system=\fRntsc|pal
.RS
Specify the video system to use.  This must match the system used by
DVswitch.  The default is "pal".
.RE
.TP
\fB\-d\fR, \fB\-\-rate=\fR48000|32000|44100
.RS
Specify the sample rate, in Hz.  The default is 48000.
.RE
.TP
\fB-d\fR, \fB--delay\fR=\fIDELAY\fR
.RS
Specify the time by which audio should be delayed, in seconds.
The delays in live audio and video sources need to be closely
matched in order that the audio and video signals will be
synchronised in the output of DVswitch.  The default is 0.2.
.RE
.SH AUTHOR
Ben Hutchings <ben@decadent.org.uk>.
.SH SEE ALSO
.UR "http://\:www.alsa-project.org/\:alsa-doc/\:alsa-lib/\:pcm.html"
ALSA PCM documentation
.UE ,
/usr/share/doc/dvswitch/README


ChangeLog
dvswitch (0.9~rc2) UNRELEASED; urgency=low

  [ Wouter Verhelst ]
  * Don't try to build or install dvsource-alsa on non-Linux.

  [ Robin Gareus ]
  * fast-fades (allow fade effect 40ms..15sec)
  * ffmpeg API update (for ffmpeg-0.10)
  * dynamic A/B overlay
  * confirm application quit dialog
  * blinking status bar
  * updated GUI widget alignment, and labels
  * various updates to Debian packaging
  * various spelling and formating fixes (doc, man-pages)

 -- Wouter Verhelst <wouter@debian.org>  Wed, 28 Mar 2012 00:39:04 +0200

dvswitch (0.9~rc1) experimental; urgency=low

  [ Ben Hutchings ]
  * Added dvsource-alsa as an audio-only source
  * Create one encoding thread per processor, up to a maximum of 8
  * Add status overlay to make it more clear whether recording is enabled
  * Set display to black before a source is connected
  * Add static Makefile at top level
  * Override source frame aspect ratio in case of a mismatch, rather then
    warning about it - Alioth feature #311953
  * Fix APTs and validity flags in header block - Alioth bug #311836
  * List all needed libraries to the linker.  We were accidentally
    depending on the behaviour of GNU ld, which treats 'needed' as
    transitive by default - Debian bug #554310
  * Remove support for 44.1 kHz audio
  * Add update-po target for updating translation files
  * Fix build failure with ffmpeg 0.6
  * Fix crash at startup with ffmpeg 0.6 due to a new parameter check
  * Enable selection of Firewire DV devices by GUID, thanks to
    Manuel Virgilio
  * Implement dvsource-firewire-rtsp, to serve DV over RTSP (experimental)

  [ Carl Karsten ]
  * Add peak level indicators to VU-meter

  [ Wouter Verhelst ]
  * Add support for tally lights in dvsource-firewire/dvsource-v4l2-dv
  * Internationalize and add Dutch translation
  * Disable effects buttons until sufficient sources have connected.
    Pressing them before sources are connected would result in a crash.
  * Implement fading on an 8-bit scale
  * Disable record button if we can't record - Alioth bug #312329
  * Add tally light support to dvsource-firewire-rtsp
  * Install dvsource-firewire-rtsp into .../bin
  * Take over day-to-day maintenance from Ben

  [ Robin Gareus ]
  * Add dvsource-jack, similar to dvsource-alsa but for JACK audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 01 Mar 2009 03:16:30 +0000

dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000

dvswitch (0.6) unstable; urgency=low

  * Added some infrastructure for video effects including
    picture-in-picture mixing
  * Switched from gtkmm 2.0/2.2 to 2.4+
  * Corrected small error in video display width

src/dvsource-alsa.c
/* Copyright 2007-2009 Ben Hutchings.
 * See the file "COPYING" for licence details.
 */
/* Source that reads audio from an ALSA device and combines with black video */

#include <assert.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>

#include <asoundlib.h>

#include "config.h"
#include "dif.h"
#include "pcm.h"
#include "protocol.h"
#include "socket.h"

static struct option options[] = {
    {"host",   1, NULL, 'h'},
    {"port",   1, NULL, 'p'},
    {"system", 1, NULL, 's'},
    {"rate",   1, NULL, 'r'},
    {"delay",  1, NULL, 'd'},
    {"help",   0, NULL, 'H'},
    {NULL,     0, NULL, 0}
};

static char * mixer_host = NULL;
static char * mixer_port = NULL;

static void handle_config(const char * name, const char * value)
{
    if (strcmp(name, "MIXER_HOST") == 0)
    {
	free(mixer_host);
	mixer_host = strdup(value);
    }
    else if (strcmp(name, "MIXER_PORT") == 0)
    {
	free(mixer_port);
	mixer_port = strdup(value);
    }
}

static void usage(const char * progname)
{
    fprintf(stderr,
	    "\
Usage: %s [-h HOST] [-p PORT] [-s ntsc|pal] \\\n\
           [-r 48000|32000|44100] [-d DELAY] [DEVICE]\n",
	    progname);
}

struct transfer_params {
    snd_pcm_t *              pcm;
    snd_pcm_uframes_t        hw_frame_count;
    const struct dv_system * system;
    enum dv_sample_rate      sample_rate_code;
    snd_pcm_uframes_t        delay_size;
    int                      sock;
};


static void transfer_frames(struct transfer_params * params)
{
    static uint8_t buf[DIF_MAX_FRAME_SIZE];
    unsigned avail_count = 0;
    unsigned serial_num = 0;

    const snd_pcm_uframes_t buffer_size =
	(params->delay_size >= 2000 ? params->delay_size : 2000)
	+ params->hw_frame_count - 1;
    pcm_sample * samples =
	malloc(sizeof(pcm_sample) * PCM_CHANNELS * buffer_size);

    dv_buffer_fill_dummy(buf, params->system);

    for (;;)
    {
	unsigned frame_count =
	    params->system->audio_frame_counts[params->sample_rate_code].std_cycle[
		serial_num % params->system->audio_frame_counts[params->sample_rate_code].std_cycle_len];

	while (avail_count < params->delay_size || avail_count < frame_count)
	{
	    snd_pcm_sframes_t rc = snd_pcm_readi(params->pcm,
						 samples + PCM_CHANNELS * avail_count,
						 params->hw_frame_count);
	    if (rc < 0)
	    {
		// Recover from buffer underrun
		if (rc == -EPIPE && snd_pcm_prepare(params->pcm) == 0)
		{
		    fprintf(stderr, "WARN: Failing to keep up with audio source\n");
		    continue;
		}
		else
		{
		    fprintf(stderr, "ERROR: snd_pcm_readi: %s\n", snd_strerror(rc));
		    exit(1);
		}
	    }
	    avail_count += rc;
	}

	dv_buffer_set_audio(buf, params->sample_rate_code, frame_count, samples);

	if (write(params->sock, buf, params->system->size)
	    != (ssize_t)params->system->size)
	{
	    perror("ERROR: write");
	    exit(1);
	}

	memmove(samples, samples + PCM_CHANNELS * frame_count,
		sizeof(pcm_sample) * PCM_CHANNELS *
		(avail_count - frame_count));
	avail_count -= frame_count;
	++serial_num;
    }
}

int main(int argc, char ** argv)
{
    /* Initialise settings from configuration files. */
    dvswitch_read_config(handle_config);

    struct transfer_params params;
    char * system_name = NULL;
    long sample_rate = 48000;
    double delay = 0.2;

    /* Parse arguments. */

    int opt;
    while ((opt = getopt_long(argc, argv, "h:p:s:r:d:", options, NULL)) != -1)
    {
	switch (opt)
	{
	case 'h':
	    free(mixer_host);
	    mixer_host = strdup(optarg);
	    break;
	case 'p':
	    free(mixer_port);
	    mixer_port = strdup(optarg);
	    break;
	case 's':
	    free(system_name);
	    system_name = strdup(optarg);
	    break;
	case 'r':
	    sample_rate = strtol(optarg, NULL, 10);
	    break;
	case 'd':
	    delay = strtod(optarg, NULL);
	    break;
	case 'H': /* --help */
	    usage(argv[0]);
	    return 0;
	default:
	    usage(argv[0]);
	    return 2;
	}
    }

    if (!mixer_host || !mixer_port)
    {
	fprintf(stderr, "%s: mixer hostname and port not defined\n",
		argv[0]);
	return 2;
    }

    if (!system_name || !strcasecmp(system_name, "pal"))
    {
	params.system = &dv_system_625_50;
    }
    else if (!strcasecmp(system_name, "ntsc"))
    {
	params.system = &dv_system_525_60;
    }
    else
    {
	fprintf(stderr, "%s: invalid system name \"%s\"\n", argv[0], system_name);
	return 2;
    }

    if (sample_rate == 32000)
    {
	params.sample_rate_code = dv_sample_rate_32k;
    }
    else if (sample_rate == 48000)
    {
	params.sample_rate_code = dv_sample_rate_48k;
    }
    else
    {
	fprintf(stderr, "%s: invalid sample rate %ld\n", argv[0], sample_rate);
	return 2;
    }

    if (delay >= 0.0)
    {
	params.delay_size = delay * sample_rate;
    }
    else
    {
	fprintf(stderr, "%s: delays do not work that way!\n", argv[0]);
	return 2;
    }

    if (argc > optind + 1)
    {
	fprintf(stderr, "%s: excess argument \"%s\"\n",
		argv[0], argv[optind + 1]);
	usage(argv[0]);
	return 2;
    }


    const char * device = (argc == optind) ? "default" : argv[optind];
    int rc;

    /* Prepare to capture and connect a socket to the mixer. */

    printf("INFO: Capturing from %s\n", device);
    rc = snd_pcm_open(&params.pcm, device, SND_PCM_STREAM_CAPTURE, 0);
    if (rc < 0)
    {
	fprintf(stderr, "ERROR: snd_pcm_open: %s\n", snd_strerror(rc));
	return 1;
    }

    snd_pcm_hw_params_t * hw_params;
    snd_pcm_hw_params_alloca(&hw_params);
    rc = snd_pcm_hw_params_any(params.pcm, hw_params);
    if (rc < 0)
    {
	fprintf(stderr, "ERROR: snd_pcm_hw_params_any: %s\n", snd_strerror(rc));
	return 1;
    }
    rc = snd_pcm_hw_params_set_access(params.pcm, hw_params, SND_PCM_ACCESS_RW_INTERLEAVED);
    if (rc >= 0)
	rc = snd_pcm_hw_params_set_format(params.pcm, hw_params, SND_PCM_FORMAT_S16);
    if (rc >= 0)
	snd_pcm_hw_params_set_channels(params.pcm, hw_params, 2);
    if (rc >= 0)
	snd_pcm_hw_params_set_rate_resample(params.pcm, hw_params, 1);
    if (rc >= 0)
	snd_pcm_hw_params_set_rate(params.pcm, hw_params, sample_rate, 0);
    if (rc >= 0)
    {
	params.hw_frame_count =
	    params.system->audio_frame_counts[params.sample_rate_code].std_cycle[0];
	rc = snd_pcm_hw_params_set_period_size_near(params.pcm, hw_params,
						    &params.hw_frame_count, 0);
    }
    if (rc >= 0)
    {
	unsigned buffer_time = 250000;
	rc = snd_pcm_hw_params_set_buffer_time_near(params.pcm, hw_params,
						    &buffer_time, 0);
    }
    if (rc >= 0)
	rc = snd_pcm_hw_params(params.pcm, hw_params);
    if (rc < 0)
    {
	fprintf(stderr, "ERROR: snd_pcm_hw_params: %s\n", snd_strerror(rc));
	return 1;
    }

    printf("INFO: Connecting to %s:%s\n", mixer_host, mixer_port);
    params.sock = create_connected_socket(mixer_host, mixer_port);
    assert(params.sock >= 0); /* create_connected_socket() should handle errors */
    if (write(params.sock, GREETING_SOURCE, GREETING_SIZE) != GREETING_SIZE)
    {
	perror("ERROR: write");
	exit(1);
    }
    printf("INFO: Connected.\n");

    transfer_frames(&params);

    close(params.sock);
    snd_pcm_close(params.pcm);

    return 0;
}
